app/src/main/java/io/appium/settings/LocationTracker.java

```diff
@@ -20,64 +20,136 @@
     private static final long FAST_INTERVAL = 5000; // 5 seconds
 
+    private static final long UPDATE_INTERVAL_IN_MILLISECONDS = 1000;
+    private static final long FASTEST_UPDATE_INTERVAL_IN_MILLISECONDS =
+            UPDATE_INTERVAL_IN_MILLISECONDS / 2;
+
     private volatile LocationManager mLocationManager;
     private volatile GoogleApiClient mGoogleApiClient;
     private volatile Location mLocation;
     private volatile boolean isStarted = false;
-    private String mLocationProvider;
-
-    private static LocationTracker instance = null;
 
-    private LocationTracker() {
-    }
+    private String mGpsLocationProviderName;
+    private String mNetworkLocationProviderName;
 
-    public synchronized static LocationTracker getInstance() {
-        if (instance == null) {
-            instance = new LocationTracker();
-        }
-        return instance;
-    }
-
-    @Override
-    public void onLocationChanged(Location location) {
+    private synchronized void updateLocation(Location location){
         if (location != null) {
             Log.d(TAG, "Got an updated location");
             mLocation = location;
         }
     }
+    private final com.google.android.gms.location.LocationListener googleApiLocationListener = new com.google.android.gms.location.LocationListener(){
+        @Override
+        public void onLocationChanged(Location location) {
+            updateLocation(location);
+        }
+    };
+    private final LocationListener gpsLocationListener =new LocationListener(){
+
+        @Override
+        public void onStatusChanged(String provider, int status, Bundle extras) {
+            switch (status) {
+                case LocationProvider.AVAILABLE:
+                    Log.d(TAG,"GPS available again\n");
+                    break;
+                case LocationProvider.OUT_OF_SERVICE:
+                    Log.d(TAG,"GPS out of service\n");
+                    break;
+                case LocationProvider.TEMPORARILY_UNAVAILABLE:
+                    Log.d(TAG,"GPS temporarily unavailable\n");
+                    break;
+            }
+        }
 
-    @Override
-    public void onProviderDisabled(String provider) {
-    }
+        @Override
+        public void onProviderEnabled(String provider) {
+            Log.d(TAG, "GPS Provider Enabled\n");
+        }
 
-    @Override
-    public void onProviderEnabled(String provider) {
+        @Override
+        public void onProviderDisabled(String provider) {
+            Log.d(TAG,"GPS Provider Disabled\n");
+        }
+
+        @Override
+        public void onLocationChanged(Location location) {
+            updateLocation(location);
+            Log.d(TAG,"New GPS location: "
+                    + String.format("%9.6f", location.getLatitude()) + ", "
+                    + String.format("%9.6f", location.getLongitude()) + "\n");
+        }
+    };
+    private final LocationListener networkLocationListener = new LocationListener(){
```

Comment:
These two classes are very similar. Please unify them to avoid unnecessary duplication